gdal raster tile: improve a bit the efficiency of multithreading …

rouault opened the pull request:
…by reducing the number of jobs to approximatively the number of threads

Fixes #12661

rouault commented:
It is strange that you are running out of file descriptors at 64 threads. What is your input: a GeoTIFF, a VRT of many sources ?

jcphill commented:
> It is strange that you are running out of file descriptors at 64 threads. What is your input: a GeoTIFF, a VRT of many sources ?

Yes, my input is a 631800 by 270000 VRT with 3631 GTiff sources of 5400 by 5400. I did set GDAL_MAX_DATASET_POOL_SIZE=800. I don't know how that is enforced internally but if threads are waiting for file descriptors to free up that contention could explain the futex calls.

jcphill commented:
Actually, those errors didn't happen until it started reading the level-13 tiles to generate the level-12 tiles, so it was done reading from the input VRT at that point and each thread should only need to have one tile open for reading and one tile open for writing at a time.

jcphill commented:
FYI, I got the "IReadBlock failed at X offset 14, Y offset 57: File /tmp/fist_seed_src_tiles/cost/13/1265/2852.png exists but cannot be opened with PNG driver" error again but file itself can be read by gdalinfo so it looks like a reading problem.

rouault commented:
> Actually, those errors didn't happen until it started reading the level-13 tiles to generate the level-12 tiles

I've added a commit in that PR to reduce the number of cached source tiles per thread from 64 to an optimal number (2 for common tiling schemes with consecutive zoom factors with  resolution varying by a factor of 2), and potentially reducing the number of threads if it sees that the number of remaining available file descriptors cannot satisfy that

jcphill commented:
With those changes my 64-core run succeeded, but only 2.3 times faster than the 8-core run with a 25x increase in system time.

8 cores:

> 0...10...20...30...40...50...60...70...80...90...100 - done in 02:47:14.
> 
> real    167m16.608s
> user    731m13.690s
> sys     103m56.294s
>
> 37% idle time
> 8% system time

64 cores:

> 0...10...20...30...40...50...60...70...80...90...100 - done in 01:12:13.
> 
> real    72m15.368s
> user    786m50.378s
> sys     2597m35.397s
>
> 27% idle time
> 56% system time

rouault commented:
@jcphill I realized that one of my yesterday change regarding sizing of cached source tiles when generating overviews was incorrect and would lead to tiles of the Z+1 zoom level being re-read many times when generating zoom Z. Now fixed.
I've also added in the last commit a --parallel-method=spawn switch that uses fork+exec instead of threading. For now I've limited that implementation to the maximum zoom level. It could potentially be extended to overview tiles, if it brings a benefit. With my current modest sized dataset, I don't see any significant different between fork+exec and spawn

jcphill commented:
Getting ready to test --parallel-method=spawn but wanted to report this error on 128 cores with the VRT sources aggregated into 50 horizontal stripes. This is on a host with 256 GB of memory and GDAL_CACHEMAX=90GB. strace indicated that each thread was iterating along a pair of columns rather than along rows. Tile indices at this lever are from 13/1251/2778 to 13/2582/3533.

> 0...10..ERROR 1: tile: GetBlockRef failed at X block offset 402, Y block offset 15: /code/gdal/gcore/gdalrasterblock.cpp, 1073: cannot allocate 262144 bytes

rouault commented:
> wanted to report this error on 128 cores with the VRT sources aggregated into 50 horizontal stripes

this error is with threading, not spawing right ?  It is very unlikely you need GDAL_CACHEMAX to be so high. Having a too high value can cause more work in shared structures.

For common understanding, would you mind sharing your exact command line, relevant env variables and output of gdalinfo on your input dataset?

jcphill commented:
> > wanted to report this error on 128 cores with the VRT sources aggregated into 50 horizontal stripes
> 
> this error is with threading, not spawing right ? It is very unlikely you need GDAL_CACHEMAX to be so high. Having a too high value can cause more work in shared structures.
> 
> For common understanding, would you mind sharing your exact command line, relevant env variables and output of gdalinfo on your input dataset?

`apptainer exec --cleanenv --env GDAL_MAX_DATASET_POOL_SIZE=800 --env GDAL_CACHEMAX=90GB $IMG gdal raster tile --convention xyz --min-zoom 2 --max-zoom 13 --progress -j 128 /tmp/striped/color.vrt /tmp/tiles`

gdalinfo color.vrt 
```Driver: VRT/Virtual Raster
Files: color.vrt
       orig.vrt
Size is 631800, 270000
Coordinate System is:
GEOGCRS["NAD83",
    DATUM["North American Datum 1983",
        ELLIPSOID["GRS 1980",6378137,298.257222101004,
            LENGTHUNIT["metre",1]]],
    PRIMEM["Greenwich",0,
        ANGLEUNIT["degree",0.0174532925199433]],
    CS[ellipsoidal,2],
        AXIS["geodetic latitude (Lat)",north,
            ORDER[1],
            ANGLEUNIT["degree",0.0174532925199433]],
        AXIS["geodetic longitude (Lon)",east,
            ORDER[2],
            ANGLEUNIT["degree",0.0174532925199433]],
    ID["EPSG",4269]]
Data axis to CRS axis mapping: 2,1
Origin = (-125.000000000000000,49.499999999524533)
Pixel Size = (0.000092592592681,-0.000092592592681)
Corner Coordinates:
Upper Left  (-125.0000000,  49.5000000) (125d 0' 0.00"W, 49d30' 0.00"N)
Lower Left  (-125.0000000,  24.5000000) (125d 0' 0.00"W, 24d30' 0.00"N)
Upper Right ( -66.4999999,  49.5000000) ( 66d30' 0.00"W, 49d30' 0.00"N)
Lower Right ( -66.4999999,  24.5000000) ( 66d30' 0.00"W, 24d30' 0.00"N)
Center      ( -95.7500000,  37.0000000) ( 95d45' 0.00"W, 37d 0' 0.00"N)
Band 1 Block=128x128 Type=Byte, ColorInterp=Red
  Mask Flags: PER_DATASET ALPHA 
Band 2 Block=128x128 Type=Byte, ColorInterp=Green
  Mask Flags: PER_DATASET ALPHA 
Band 3 Block=128x128 Type=Byte, ColorInterp=Blue
  Mask Flags: PER_DATASET ALPHA 
Band 4 Block=128x128 Type=Byte, ColorInterp=Alpha
```

gdalinfo orig.vrt 
```Driver: VRT/Virtual Raster
Files: orig.vrt
       orig.10.tif
       orig.11.tif
       orig.12.tif
       orig.13.tif
       orig.14.tif
       orig.15.tif
       orig.16.tif
       orig.17.tif
       orig.18.tif
       orig.19.tif
       orig.1.tif
       orig.20.tif
       orig.21.tif
       orig.22.tif
       orig.23.tif
       orig.24.tif
       orig.25.tif
       orig.26.tif
       orig.27.tif
       orig.28.tif
       orig.29.tif
       orig.2.tif
       orig.30.tif
       orig.31.tif
       orig.32.tif
       orig.33.tif
       orig.34.tif
       orig.35.tif
       orig.36.tif
       orig.37.tif
       orig.38.tif
       orig.39.tif
       orig.3.tif
       orig.40.tif
       orig.41.tif
       orig.42.tif
       orig.43.tif
       orig.44.tif
       orig.45.tif
       orig.46.tif
       orig.47.tif
       orig.48.tif
       orig.49.tif
       orig.4.tif
       orig.50.tif
       orig.5.tif
       orig.6.tif
       orig.7.tif
       orig.8.tif
       orig.9.tif
Size is 631800, 270000
Coordinate System is:
GEOGCRS["NAD83",
    DATUM["North American Datum 1983",
        ELLIPSOID["GRS 1980",6378137,298.257222101004,
            LENGTHUNIT["metre",1]]],
    PRIMEM["Greenwich",0,
        ANGLEUNIT["degree",0.0174532925199433]],
    CS[ellipsoidal,2],
        AXIS["geodetic latitude (Lat)",north,
            ORDER[1],
            ANGLEUNIT["degree",0.0174532925199433]],
        AXIS["geodetic longitude (Lon)",east,
            ORDER[2],
            ANGLEUNIT["degree",0.0174532925199433]],
    ID["EPSG",4269]]
Data axis to CRS axis mapping: 2,1
Origin = (-125.000000000000000,49.499999999524533)
Pixel Size = (0.000092592592681,-0.000092592592681)
Corner Coordinates:
Upper Left  (-125.0000000,  49.5000000) (125d 0' 0.00"W, 49d30' 0.00"N)
Lower Left  (-125.0000000,  24.5000000) (125d 0' 0.00"W, 24d30' 0.00"N)
Upper Right ( -66.4999999,  49.5000000) ( 66d30' 0.00"W, 49d30' 0.00"N)
Lower Right ( -66.4999999,  24.5000000) ( 66d30' 0.00"W, 24d30' 0.00"N)
Center      ( -95.7500000,  37.0000000) ( 95d45' 0.00"W, 37d 0' 0.00"N)
Band 1 Block=128x128 Type=Float32, ColorInterp=Gray
  NoData Value=inf
```

head -22 orig.vrt 
```<VRTDataset rasterXSize="631800" rasterYSize="270000">
  <SRS dataAxisToSRSAxisMapping="2,1">GEOGCS["NAD83",DATUM["North_American_Datum_1983",SPHEROID["GRS 1980",6378137,298.257222101004,AUTHORITY["EPSG","7019"]],AUTHORITY["EPSG","6269"]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]],AXIS["Latitude",NORTH],AXIS["Longitude",EAST],AUTHORITY["EPSG","4269"]]</SRS>
  <GeoTransform> -1.2500000000000000e+02,  9.2592592680929993e-05,  0.0000000000000000e+00,  4.9499999999524533e+01,  0.0000000000000000e+00, -9.2592592680642354e-05</GeoTransform>
  <VRTRasterBand dataType="Float32" band="1">
    <NoDataValue>inf</NoDataValue>
    <ColorInterp>Gray</ColorInterp>
    <ComplexSource>
      <SourceFilename relativeToVRT="1">orig.10.tif</SourceFilename>
      <SourceBand>1</SourceBand>
      <SourceProperties RasterXSize="631800" RasterYSize="5400" DataType="Float32" BlockXSize="256" BlockYSize="256" />
      <SrcRect xOff="0" yOff="0" xSize="631800" ySize="5400" />
      <DstRect xOff="0" yOff="48600" xSize="631800" ySize="5400" />
      <NODATA>inf</NODATA>
    </ComplexSource>
    <ComplexSource>
      <SourceFilename relativeToVRT="1">orig.11.tif</SourceFilename>
      <SourceBand>1</SourceBand>
      <SourceProperties RasterXSize="626400" RasterYSize="5400" DataType="Float32" BlockXSize="256" BlockYSize="256" />
      <SrcRect xOff="0" yOff="0" xSize="626400" ySize="5400" />
      <DstRect xOff="0" yOff="54000" xSize="626400" ySize="5400" />
      <NODATA>inf</NODATA>
    </ComplexSource>
```

jcphill commented:
> Getting ready to test --parallel-method=spawn but wanted to report this error on 128 cores with the VRT sources aggregated into 50 horizontal stripes. This is on a host with 256 GB of memory and GDAL_CACHEMAX=90GB. strace indicated that each thread was iterating along a pair of columns rather than along rows. Tile indices at this level are from 13/1251/2778 to 13/2582/3533.
> 
> > 0...10..ERROR 1: tile: GetBlockRef failed at X block offset 402, Y block offset 15: /code/gdal/gcore/gdalrasterblock.cpp, 1073: cannot allocate 262144 bytes

Completed successfully with GDAL_CACHEMAX to 1GB.

elpaso commented on apps/gdalalg_raster_tile.cpp:
```diff
@@ -3206,7 +3839,27 @@ bool GDALRasterTileAlgorithm::RunImpl(GDALProgressFunc pfnProgress,
                             oResourceManager.GetErrorMsg().c_str());
             }
         }
+        else
```
this is the branch for no threads and no processes, right?

rouault replied:
yes, adding a code comment

jcphill commented:
> Spawn mode extended to overview tile generation, and automatically used if GDAL can use it and there is enough tiles per job to generate to make it worth to spawn a child process (arbitrarily set at a minimum of 100 tiles per job)

Much faster but seeing the following errors repeatably (i.e., same files reporting errors) on 128 cores. The files reporting the read errors are identical to those from successful runs.

> apptainer exec --cleanenv --env GDAL_MAX_DATASET_POOL_SIZE=800 --env GDAL_CACHEMAX=90GB gdal_fix_12661_v4.sif gdal raster tile --convention xyz --min-zoom 2 --max-zoom 13 --progress --parallel-method=spawn -j 128 color.vrt /tmp/tiles
> 0...10...20...30...40...50...60...70...80...90...ERROR 1: posix_spawnp() failed
> ERROR 1: tile: Spawning child gdal process '/usr/local/bin/gdal raster tile --config GDAL_NUM_THREADS=1 --config GDAL_CACHEMAX=754974720 --num-threads 1 --ovr-zoom-level 9 --ovr-min-x 106 --ovr-max-x 106 --ovr-min-y 197 --ovr-max-y 197 --webviewer none --progress-forked --input color.vrt --output /tmp/tiles --min-zoom 2 --max-zoom 13 --convention xyz' failed
> ERROR 1: tile:  File /tmp/tiles/10/205/395.png exists but cannot be opened with PNG driver
> ERROR 1: tile:  IReadBlock failed at X offset 49, Y offset 48: File /tmp/tiles/10/205/395.png exists but cannot be opened with PNG driver
> ERROR 1: tile: Child process '/usr/local/bin/gdal raster tile --config GDAL_NUM_THREADS=1 --config GDAL_CACHEMAX=754974720 --num-threads 1 --ovr-zoom-level 9 --ovr-min-x 103 --ovr-max-x 103 --ovr-min-y 197 --ovr-max-y 197 --webviewer none --progress-forked --input color.vrt --output /tmp/tiles --min-zoom 2 --max-zoom 13 --convention xyz' failed
> ERROR 1: tile:  File /tmp/tiles/10/209/394.png exists but cannot be opened with PNG driver
> ERROR 1: tile:  IReadBlock failed at X offset 53, Y offset 47: File /tmp/tiles/10/209/394.png exists but cannot be opened with PNG driver
> ERROR 1: tile: Child process '/usr/local/bin/gdal raster tile --config GDAL_NUM_THREADS=1 --config GDAL_CACHEMAX=754974720 --num-threads 1 --ovr-zoom-level 9 --ovr-min-x 104 --ovr-max-x 104 --ovr-min-y 197 --ovr-max-y 197 --webviewer none --progress-forked --input color.vrt --output /tmp/tiles --min-zoom 2 --max-zoom 13 --convention xyz' failed
> ERROR 1: tile:  File /tmp/tiles/10/209/394.png exists but cannot be opened with PNG driver
> ERROR 1: tile:  IReadBlock failed at X offset 53, Y offset 47: File /tmp/tiles/10/209/394.png exists but cannot be opened with PNG driver
> ERROR 1: tile: Child process '/usr/local/bin/gdal raster tile --config GDAL_NUM_THREADS=1 --config GDAL_CACHEMAX=754974720 --num-threads 1 --ovr-zoom-level 9 --ovr-min-x 105 --ovr-max-x 105 --ovr-min-y 197 --ovr-max-y 197 --webviewer none --progress-forked --input color.vrt --output /tmp/tiles --min-zoom 2 --max-zoom 13 --convention xyz' failed

rouault commented:
> Indeed it is.

excellent. What is the runtime of your test case ? (did you compare it with gdal2tiles ?)

jcphill commented:
> excellent. What is the runtime of your test case ? (did you compare it with gdal2tiles ?)

gdal raster tile --parallel-method=spawn -j 128:

> 0...10...20...30...40...50...60...70...80...90...100 - done in 00:12:38.
> real    12m42.712s
> user    767m57.402s
> sys     68m26.785s

On a different, 64-core machine where gdal2tiles is working (long story)...

gdal raster tile --parallel-method=spawn -j 64:

> 0...10...20...30...40...50...60...70...80...90...100 - done in 00:20:19.
> real	20m21.979s
> user	685m20.837s
> sys	64m46.123s

gdal2tiles.py --processes 64:

> 90% done at 45 minutes, so around 50 minutes.

So a factor of 2.5 faster than gdal2tiles.

jcphill commented on apps/gdalalg_raster_tile.cpp:
```diff
@@ -2894,9 +3630,66 @@ bool GDALRasterTileAlgorithm::RunImpl(GDALProgressFunc pfnProgress,
+    }
+    else if (m_numThreads > 1 &&
+             ((m_parallelMethod.empty() &&
+               nBaseTiles / m_numThreads > THRESHOLD_TILES_PER_JOB &&
```
This is the wrong criterion for switching from threads to spawning processes. The degree of parallelism numThreads should be limited to ceil(nXXXTiles / THRESHOLD_TILES_PER_JOB) and spawning used if numThreads > 8 or so. With the current method, if I ask for -j 128 but there are only 12,000 tiles I'll get 128 threads (at least that's what it looks like).

jcphill commented on apps/gdalalg_raster_tile.cpp:
```diff
@@ -3668,9 +3684,11 @@ bool GDALRasterTileAlgorithm::RunImpl(GDALProgressFunc pfnProgress,
-               nBaseTiles / m_numThreads > THRESHOLD_TILES_PER_JOB &&
+               m_numThreads >= THRESHOLD_MIN_THREADS_FOR_SPAWN &&
+               nBaseTiles >=
+                   THRESHOLD_MIN_THREADS_FOR_SPAWN * THRESHOLD_TILES_PER_JOB &&
```
This has the same problem as before. Think of 8 as MAX_THREADS_FOR_THREAD and never launch more than that number of threads without switching to spawn.

rouault replied:
not totally sure to follow you. I've added a commit that limit the maximum number of threads (for the threading use case) to min(number_of_threads_specified_by_user_or_all_available, number_of_tiles / 100) . If that's not what you've in mind, a diff / pull request with your proposed changes could make it easier

jcphill replied:
Now that you have limited m_numThreads you can just eliminate the "nBaseTiles >= ..." condition and it's good.

jcphill commented:
@rouault I pulled and tested master and I get a repeatable segfault when starting overview tiles if I do not specify --parallel-method=spawn even though it should definitely still be using spawn at that stage. No idea what could cause this. Building debug now.

jcphill commented:
Caught it:

```
0...10...20...30...40...50...60...70.                - estimated remaining time: 00:16:03          
Program received signal SIGSEGV, Segmentation fault.
GDALRasterTileAlgorithm::IsCompatibleOfSpawn (this=0x7ffff407d010, 
    pszErrorMsg=@0x7fffffffd778: 0x7ffff782cf4a "") at /code/gdal/apps/gdalalg_raster_tile.cpp:2302
2302        auto poSrcDriver = m_poSrcDS->GetDriver();
(gdb) where
#0  GDALRasterTileAlgorithm::IsCompatibleOfSpawn (this=0x7ffff407d010, 
    pszErrorMsg=@0x7fffffffd778: 0x7ffff782cf4a "") at /code/gdal/apps/gdalalg_raster_tile.cpp:2302
#1  0x00007ffff73e4a55 in operator() (__closure=0x7fffffffd998)
    at /code/gdal/apps/gdalalg_raster_tile.cpp:3676
#2  0x00007ffff73eaa87 in GDALRasterTileAlgorithm::RunImpl (this=0x7ffff407d010, 
    pfnProgress=0x7ffff5e3e1f7 <GDALTermProgress(double, char const*, void*)>, pProgressData=0x0)
    at /code/gdal/apps/gdalalg_raster_tile.cpp:3999
#3  0x00007ffff70a48d7 in GDALAlgorithm::Run (this=0x7ffff407d010, 
    pfnProgress=0x7ffff5e3e1f7 <GDALTermProgress(double, char const*, void*)>, pProgressData=0x0)
    at /code/gdal/gcore/gdalalgorithm.cpp:5067
#4  0x00007ffff70a470f in GDALAlgorithm::Run (this=0x7ffff4095a90, 
    pfnProgress=0x7ffff5e3e1f7 <GDALTermProgress(double, char const*, void*)>, pProgressData=0x0)
    at /code/gdal/gcore/gdalalgorithm.cpp:5027
#5  0x00007ffff70a470f in GDALAlgorithm::Run (this=0x7ffff4134a90, 
    pfnProgress=0x7ffff5e3e1f7 <GDALTermProgress(double, char const*, void*)>, pProgressData=0x0)
    at /code/gdal/gcore/gdalalgorithm.cpp:5027
#6  0x0000555555583be6 in main (argc=14, argv=0x7ffff408a790) at /code/gdal/apps/gdal.cpp:220
(gdb) list
2297    /************************************************************************/
2298
2299    bool GDALRasterTileAlgorithm::IsCompatibleOfSpawn(const char *&pszErrorMsg)
2300    {
2301        pszErrorMsg = "";
2302        auto poSrcDriver = m_poSrcDS->GetDriver();
2303        if (m_poSrcDS->GetDescription()[0] == 0 || !poSrcDriver ||
2304            EQUAL(poSrcDriver->GetDescription(), "MEM"))
2305        {
2306            pszErrorMsg = "Unnamed or memory dataset sources are not supported "
(gdb) print m_poSrcDS
$1 = (GDALDataset *) 0x0
```

jcphill commented:
Set to null at https://github.com/OSGeo/gdal/blob/b5d2a80d4bf93493a464c18623b7acc74e034fac/apps/gdalalg_raster_tile.cpp#L3944-L3948

There is no need to check IsCompatibleOfSpawn on the source for the overview tiles.

jcphill commented:
> fixed per [30a5145](https://github.com/OSGeo/gdal/commit/30a5145e87646158e43df515f2b6cd4dff91e679)

It works now. Thanks for getting this wrapped up quickly!

rouault commented:
> but it looks like these commits are in Master two weeks ago, so does that mean they are down stream in 3.11.3?

master is 3.12dev, so no, they are not in 3.11.3 as they haven't being backported to the stable 3.11 branch
